SliceViewer bug fixes

zjmorgan opened the pull request:
**Description of work.**

General improvements to sliceviewer.

1. Default slicing to zero if available. 
3. Choose log scale and zoom into an all zero slice to ensure no exceptions are thrown.
4. Add line plots and zoom in. Make sure the lineplots follow accurately reflect the data.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

<!-- Instructions for testing. -->

Use this script to generate fake data

# MDHisto

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

signal = np.linspace(0,1,1331)
error = np.sqrt(signal)

signal -= 0.5

signal[signal>0.25] = np.nan

ws = CreateMDHistoWorkspace(Dimensionality=3,
                            Extents='-3,3,-10,10,-5,5',
                            SignalInput=signal,
                            ErrorInput=error,
                            NumberOfBins='11,11,11',
                            Names='[H,0,0],[0,K,0],[0,0,L]',
                            Units='r.l.u.,r.l.u.,r.l.u.',
                            Frames='HKL,HKL,HKL')

SetMDFrame('ws', MDFrame='HKL', Axes=[0,1,2])

AddSampleLog('ws', LogName='sample')
ws.getExperimentInfo(0).run().addProperty('W_MATRIX', list(np.eye(3).flatten()), True)

SetUB('ws', a=5, b=5, c=7, alpha=90, beta=90, gamma=120)
```

Open SliceViewer: verify the zero slice is chosen
![image](https://user-images.githubusercontent.com/13754794/189913787-4a76b5d1-04cc-41d6-9540-460bf350d7d5.png)

Switch to nonortho and LogScale: 
![image](https://user-images.githubusercontent.com/13754794/189913887-3c305124-ff04-4961-ade2-9945ce39c8a2.png)

Zoom into the white space area (all zeros): ensure no exceptions
![image](https://user-images.githubusercontent.com/13754794/189913928-ad402d06-3393-4da3-b55f-0de3364c4e5f.png)

Zoom into an area with data and turn on lineplots: verify the lineplot is correct
![image](https://user-images.githubusercontent.com/13754794/189917376-7dac3c7a-820b-4a47-b95b-251b81540c64.png)


# MDEvent

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateMDWorkspace(Dimensions='3', 
                       EventType='MDEvent',
                       Extents='-10,10,-10,10,-10,10',
                       Names='[H,0,0],[0,K,0],[0,0,L]',
                       Units='r.l.u.,r.l.u.,r.l.u.')

FakeMDEventData(ws, UniformParams="1000000")

SetMDFrame('ws', MDFrame='HKL', Axes=[0,1,2])

AddSampleLog('ws', LogName='sample')
ws.getExperimentInfo(0).run().addProperty('W_MATRIX', list(np.eye(3).flatten()), True)

SetUB('ws', a=5, b=5, c=7, alpha=90, beta=90, gamma=120)
```

Open SliceViewer: verify the zero slice is chosen
![image](https://user-images.githubusercontent.com/13754794/189914091-4de92aa5-9e8a-4039-8019-514567fd421a.png)
Switch to nonortho and LogScale: 

Switch to nonortho and LogScale: 
![image](https://user-images.githubusercontent.com/13754794/189914158-a853c311-37b5-462a-9a6a-da396e26cd0f.png)

Zoom into the white space area (all zeros): ensure no exceptions
![image](https://user-images.githubusercontent.com/13754794/189914187-21724581-24b0-4373-99dc-0ab8c3875d50.png)

Zoom into an area with data and turn on lineplots: verify the lineplot is correct
![image](https://user-images.githubusercontent.com/13754794/189917036-19e99a06-b4bb-44ef-803a-c5af97ee644e.png)


<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

zjmorgan committed:
Adding zero slicing default

zjmorgan committed:
Colorbar fix for log-scaling

zjmorgan committed:
Code cleanup

zjmorgan committed:
Fix zoom scaling on attached lineplots

zjmorgan committed:
Clean up history

martyngigg commented:
Are these critical bug fixes for 6.5 now that the freeze has passed?

RichardWaiteSTFC commented:
> Are these critical bug fixes for 6.5 now that the freeze has passed?

I think this PR does the following 3 things
1. I think this does fix a bug causing an exception when zooming into regions with no data (I swear we fix this every time... if there isn't already a test for this it would be good to have one). Because it causes an exception it would be good to fix in this release.
2. There is a nice to have of the initial slicepoint being 0, rather than the minimum possible value (from the code it looks like this is only if the spinbox max >0 and min < 0) - alternatively we could just start the slice at the halfway point (but for HKL data such a slice would be unlikely to contain any peaks). This is technically a new feature but I think is very minor and safe so could go in the release as well (but happy either way).
3. There are also some changes to 'fix' line-plot zooming (weirdly by removing some autoscale commands) - though I can't reproduce an issue on the nightly on IDAaaS. This sounds like a bug that we could fix in this release as well - though I don't think it causes an exception so happy either way - but it would be good to get some clarification on the original issue/problem.

FYI there's some confusion in the commit history regarding when to allow non-axis cut, but this looks unchanged and isn't mentioned in the release docs.

RichardWaiteSTFC commented on qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py:
```diff
@@ -314,17 +314,14 @@ def _calculate_clim(self) -> tuple:
-            except (AttributeError, IndexError):
-                data = data[np.nonzero(data)] if log_normalisation else data
-                return np.nanmin(data), np.nanmax(data)
+            if log_normalisation:
```
This check works well, it would be good to put a test in `test_colorbaqr.py` that supplies NaN data to stop a regression.

From the old code it looks like we would expect non-finite data to be masked in `data.mask`?
For image objects `QuadMeshWrapper` and `ModestImage` with the property `get_array_clipped_to_bounds()` (which I thought was used in non-ortho meshes) it looks like the `data.mask` is produced by the call to `ma.masked_invalid` which should do the job, but this isn't called in the fallback `get_array()` (e.g. `SamplingImage`). I don't think we can mask the data returned from `get_array()` as it will mess up the reshape in `get_array_clipped_to_bounds()` - so this is a good solution. If it isn't used elsewhere perhaps we no longer need to mask invalid data in `get_array_clipped_to_bounds()`? But maybe safer to keep as is?

RichardWaiteSTFC commented on qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/lineplots.py:
```diff
@@ -127,9 +127,7 @@ def update_line_plot_limits(self):
         """
         # ensure plot labels are in sync with main axes
         self._axx.relim()
-        self._axx.autoscale(axis='y')
```
How were these causing a problem? It seems to behave fine on a nightly version for me.

zjmorgan replied:
It seems the issue is a bit phantom. I went back over the older version and it also seems to work in most cases. There are a few instances where I can trick the extents to be messed up but I cannot currently reliably reproduce it. So for this PR, I will remove these changes until I can better characterize the problem.

RichardWaiteSTFC reviewed (changes requested):
Followed testing instructions - no errors found and everything seems to work well. I like setting the initial slicepoint to 0 (if in range) - this is more relevant for HKL data, but I think it is a better default. it also works when you change dimension/viewing axes as well.

Just a couple of things:
- Can you add a test  in `test_colorbar.py` for NaN values (probably for all colorbar scales types)
- Can you tidy up the commit history (e.g. with interactive rebase) to remove references to the non-axis aligned cut tool that wasn't changed.

With this fix it might no longer be necessary to call `np.ma.masked_array` on the return of `get_array_clipped_to_bounds` (as from a quick grep I don't think the `data.mask` is used elsewhere) but perhaps safer to keep as is.

zjmorgan committed:
Adding much needed colorbar tests

zjmorgan committed:
Updated tests for all options

zjmorgan commented:
> * Can you add a test  in `test_colorbar.py` for NaN values (probably for all colorbar scales types)
Good suggestion. I added some to the unit tests for zero and NaN values. I believe the previous fix only worked for NaN values only.

> * Can you tidy up the commit history (e.g. with interactive rebase) to remove references to the non-axis aligned cut tool that wasn't changed.
I removed those commits.

> With this fix it might no longer be necessary to call `np.ma.masked_array` on the return of `get_array_clipped_to_bounds` (as from a quick grep I don't think the `data.mask` is used elsewhere) but perhaps safer to keep as is.
I think it is safe to keep it as is. I am not sure if `get_array_clipped_to_bounds` does for all NaN or zero slices.

zjmorgan committed:
Meant to mean zero slices

zjmorgan committed:
Remove fixes for phantom lineplot extents

RichardWaiteSTFC committed:
Remove temp file

RichardWaiteSTFC reviewed (approved):
Have removed a temporary file I think was committed by mistake -  test passes locally for me so approving! Thanks for adding the tests, and I agree it's best to leave the line plot axes scaling bug till the next release if we can't reproduce it.
